Assets/_BForBoss/_Character/Scripts/PlayerWallRunBehaviour.cs

```diff
@@ -0,0 +1,232 @@
+            velocity = velocity.dot(alongWall) * alongWall;
+            velocity += Vector3.down * _wallGravityDownForce * Time.deltaTime;
+            _baseCharacter.SetVelocity(velocity);
+        }
+
+        public bool CalcJumpVelocity(out Vector3 velocity)
+        {
+            velocity = Vector3.zero;
+            if (_isWallRunning)
+            {
+                velocity = _lastWallRunNormal * _baseCharacter.jumpImpulse + Vector3.up;
+            }
+            return _isWallRunning;
+        }
+
+        private void HandleEyePivotRotation()
+        {
+            var rotation = _fpsCharacter.eyePivot.localEulerAngles;
+            rotation.z = GetCameraRoll();
+            _fpsCharacter.eyePivot.localEulerAngles = rotation;
+        }
+
+        private void WallRun(RaycastHit wall)
+        {
+            _lastWallRunNormal = wall.normal;
+            _lastWallRunPosition = wall.point;
+            if (!_isWallRunning)
+            {
+                _isWallRunning = true;
+                _baseMaxSpeed = _baseCharacter.maxWalkSpeed;
+                _baseCharacter.maxWalkSpeed *= _speedMultiplier;
+                _timeSinceWallAttach = 0f;
+                _timeSinceWallDetach = 0f;
+            }
+        }
+
+        private bool CanWallRun()
+        {
+            if (_baseCharacter.IsOnGround()) return false;
+
+            if (_baseCharacter.IsJumping())
+            {
+                _currentJumpDuration += Time.deltaTime;
+                if (_currentJumpDuration < _minJumpDuration) return false;
+            }
+
+            return !Physics.Raycast(ChildTransform.position, Vector3.down, _minHeight);
+        }
+
+        private void StopWallRunning()
+        {
+            if (!_isWallRunning) return;
+            _isWallRunning = false;
+            _baseCharacter.maxWalkSpeed = _baseMaxSpeed;
+            _timeSinceWallAttach = 0f;
+            _timeSinceWallDetach = 0f;
+        }
+
+        private float CalculateSide()
+        {
+            if (_isWallRunning)
+            {
+                Vector3 heading = _lastWallRunPosition - transform.position;
+                Vector3 perp = Vector3.Cross(ChildTransform.forward, heading);
+                float dir = Vector3.Dot(perp, ChildTransform.up);
+                return dir;
+            }
+            return 0;
+        }
+
+        private float GetCameraRoll()
+        {
+            float dir = CalculateSide();
+            float cameraAngle = _baseCharacter.cameraTransform.eulerAngles.z;
+            float targetAngle = 0;
+            if (dir != 0)
+            {
+                targetAngle = Mathf.Sign(dir) * _maxCameraAngleRoll;
+            }
+            return Mathf.LerpAngle(cameraAngle, targetAngle, Mathf.Max(_timeSinceWallAttach, _timeSinceWallDetach) / _cameraRotateDuration);
+        }
+
+        private static bool GetSmallestRaycastHit(RaycastHit[] array, out RaycastHit smallest)
```

Comment:
Rename this to `GetSmallestRaycastHitIfValid`

so when placed into an if statement, it is more sound.

Reply:
It's just personal general coding habits, by default I set it to non-static. Looking through the method it seems fine in this case. 
and just read through when other's set to static.
https://stackoverflow.com/questions/2713601/when-to-make-a-method-static